### PR TITLE
FS: use optional<string> to represent non-existent paths

### PR DESCRIPTION
**Pull request by Vultraz**
Most `filesystem::` functions that returned an empty string for an invalid path now return `optional<string>` . There are a couple areas (mostly with the schema files) where I'm not sure if I should use `value_or("")`, since it's possible with `value` we get the wrong exception thrown for an invalid path.

### CODE REVIEW

**Comment by CelticMinstrel** [comment]
I don't really have much to say here. It's a very large yet very boring change.

File: src/scripting/lua_fileops.cpp [hunk 1]
```diff
@@ -70,7 +70,7 @@ static int intf_have_asset(lua_State* L)
 static int intf_resolve_asset(lua_State* L)
 {
 	std::string type = luaL_checkstring(L, 1), name = luaL_checkstring(L, 2);
-	lua_push(L, filesystem::get_independent_binary_file_path(type, name));
+	lua_push(L, filesystem::get_independent_binary_file_path(type, name).value_or(""));
```
[the remark this answers is not in the document]
Comment (Vultraz) [reply src/scripting/lua_fileops.cpp:73]: How is that done?

[the remark this answers is not in the document]
Comment (CelticMinstrel) [reply src/scripting/lua_fileops.cpp:73]: Leaving off `.value_or("")` theoretically _should_ just do the right thing, but I'm not sure if `lua_push` has a specialization for optionals…

**Comment by soliton-** [comment]
Plenty of unchecked `value()` calls as far as I can tell. Does not look like this was ready to merge.

I'm also not a fan of the tendency to use auto everywhere you can. It makes the code less readable.

**Comment by Pentarctagon** [comment]
Probably all (or at least a bunch) of these checks should be `.value_or("")` to keep current functionality?

**Pull request merged**